cmdundo.go

```diff
@@ -25,37 +30,155 @@ var (
-	if err := cmdReplace1(args); err != nil {
+func cmdUndo(_ *Command, args []string) {
+	if err := cmdUndo1(args); err != nil {
 		errorf("%v", err)
 	}
 }
 
-func cmdReplace1(modules []string) error {
-	if len(modules) == 0 {
+func cmdUndo1(modules []string) error {
+	modFile, err := goModInfo()
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	modMap := make(map[string]bool)
+	if len(modules) > 0 {
+		for _, m := range modules {
+			modMap[m] = true
+		}
+	} else {
 		// With no modules specified, we un-gohack all modules
 		// we can find with local directory info in the go.mod file.
-		m, err := goModInfo()
-		if err != nil {
-			return errors.Wrap(err)
-		}
-		for _, r := range m.Replace {
+		for _, r := range modFile.Replace {
 			if r.Old.Version == "" && r.New.Version == "" {
+				modMap[r.Old.Path] = true
 				modules = append(modules, r.Old.Path)
 			}
 		}
 	}
-	// TODO get information from go.mod to make sure we're
-	// dropping a directory replace, not some other kind of replacement.
-	args := []string{"mod", "edit"}
-	for _, m := range modules {
-		args = append(args, "-dropreplace="+m)
+	drop := make(map[string]bool)
+	for _, r := range modFile.Replace {
+		if !modMap[r.Old.Path] || r.Old.Version != "" || r.New.Version != "" {
+			continue
+		}
+		// Found a replacement to drop.
+		comments := r.Syntax.Comments
+		if len(comments.Suffix) == 0 {
+			// No comment; we can just drop it.
+			drop[r.Old.Path] = true
+			delete(modMap, r.Old.Path)
+			continue
+		}
+		prevReplace := splitWasComment(comments.Suffix[0].Token)
+		if prevReplace != nil && prevReplace.Old.Path == r.Old.Path {
+			// We're popping the old replace statement.
+			if !r.Syntax.InBlock {
+				// When we're not in a block, we don't need the "replace" token.
+				prevReplace.Syntax.Token = prevReplace.Syntax.Token[1:]
+			}
+			// Preserve any before and after comments.
+			prevReplace.Syntax.Before = r.Syntax.Before
+			prevReplace.Syntax.After = r.Syntax.After
+			*r = *prevReplace
+			delete(modMap, r.Old.Path)
+		} else {
+			// It's not a "was" comment. Just remove it (after this loop so we don't
+			// interfere with the current range statement).
+			drop[r.Old.Path] = true
+			delete(modMap, r.Old.Path)
+		}
 	}
-	if _, err := runCmd(cwd, "go", args...); err != nil {
-		return errors.Notef(err, nil, "failed to remove go.mod replacements")
+	for m := range drop {
+		if err := modFile.DropReplace(m, ""); err != nil {
+			return errors.Notef(err, nil, "cannot drop replacement for %v", m)
+		}
+	}
+	failed := make([]string, 0, len(modMap))
+	for m := range modMap {
+		failed = append(failed, m)
+	}
+	sort.Strings(failed)
+	for _, m := range failed {
+		errorf("%s not currently replaced; cannot drop", m)
+	}
+	if err := writeModFile(modFile); err != nil {
+		return errors.Wrap(err)
 	}
 	for _, m := range modules {
 		fmt.Printf("dropped %s\n", m)
 	}
 	return nil
 }
+
+func splitWasComment(s string) *modfile.Replace {
```

Comment:
why not implement this with regexp and capture groups? This kind of string handling code always looks complicated to me.

Reply:
yeah, not a bad idea. done.